<a href="https://colab.research.google.com/github/erick-huang15/Food-Sales-Predictions/blob/main/Food_Sale_Prediction_Pt5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.compose import make_column_selector, make_column_transformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.pipeline import make_pipeline
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LinearRegression
from sklearn.dummy import DummyRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import make_column_selector, make_column_transformer
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import BaggingRegressor

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
path = '/content/drive/MyDrive/CD: Project/sales_predictions.csv'
data = pd.read_csv(path)
data.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [4]:
data.isna().sum() # this is to check the missing values in the dataset

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

In [5]:
data.info() # this is to check the types of data in the datasets

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


In [6]:
for col in data: #this is to loop through the dataset, explore, and see what changes are necessary
  print(data[col].value_counts())
  print('\n\n')

FDW13    10
FDG33    10
NCY18     9
FDD38     9
DRE49     9
         ..
FDY43     1
FDQ60     1
FDO33     1
DRF48     1
FDC23     1
Name: Item_Identifier, Length: 1559, dtype: int64



12.150    86
17.600    82
13.650    77
11.800    76
15.100    68
          ..
7.275      2
7.685      1
9.420      1
6.520      1
5.400      1
Name: Item_Weight, Length: 415, dtype: int64



Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64



0.000000    526
0.076975      3
0.162462      2
0.076841      2
0.073562      2
           ... 
0.013957      1
0.110460      1
0.124646      1
0.054142      1
0.044878      1
Name: Item_Visibility, Length: 7880, dtype: int64



Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat 

Medium    2793 
Small     2388
High       932
Name: Outlet_Size, dtype: int64 -> ordinal encode (check)



Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64 -> change LF, Reg (check)


Tier 3    3350
Tier 2    2785
Tier 1    2388
Name: Outlet_Location_Type, dtype: int64



In [7]:
replace_fc = {'LF':'Low Fat','low fat':'Low Fat','reg':'Regular'} # this is to fix the inconsistent data in fat content
data['Item_Fat_Content'] = data['Item_Fat_Content'].replace(replace_fc)
data['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

In [8]:
# this is to ordinal encode Outlet_Size to small = 1, medium = 2, high = 3
outlet_size_replace = {'High':3,'Medium':2,'Small':1}
data['Outlet_Size'].replace(outlet_size_replace,inplace=True)
data['Outlet_Size'].value_counts()

2.0    2793
1.0    2388
3.0     932
Name: Outlet_Size, dtype: int64

In [9]:
outlet_location_replace = {'Tier 3':3,'Tier 2':2,'Tier 1':1}
data['Outlet_Location_Type'].replace(outlet_location_replace,inplace=True)
data['Outlet_Location_Type'].value_counts()

3    3350
2    2785
1    2388
Name: Outlet_Location_Type, dtype: int64

In [10]:
X = data.drop(columns='Item_Outlet_Sales') # Train Test Split
y = data['Item_Outlet_Sales']
X_train, X_test, y_train, y_test = train_test_split(X,y, random_state=42)

In [22]:
# this is to instantiate colulmn selectors
category_select = make_column_selector(dtype_include='object')
numeric_select = make_column_selector(dtype_include='number')

In [35]:
# this is to impute the missing values in Item_Weight and Outlet_Size. 
# both are float and numeric data
mean_imputer = SimpleImputer(strategy='mean')
most_freq_imputer = SimpleImputer(strategy='most_frequent')

scaler = StandardScaler() # this is to scale the data so that we can make calculations or comparisons
ohe = OneHotEncoder(sparse=False,handle_unknown='ignore')

In [40]:
# this is to instantiate pipelines to make the pipeline perform tasks in a specific order
numeric_pipeline = make_pipeline(mean_imputer,scaler)
categoric_pipeline = make_pipeline(most_freq_imputer, ohe)

In [41]:
# this is to instantiate column transformer
numeric_tuple = (numeric_pipeline, numeric_select)
categoric_tuple = (categoric_pipeline, category_select)
col_transformer = make_column_transformer(numeric_tuple,categoric_tuple,remainder='passthrough')

In [49]:
# Extract the feature names from one hot encoder
cat_feature_names = col_transformer.named_transformers_['pipeline-2']\
                          .named_steps['onehotencoder']\
                          .get_feature_names_out(category_select(X_train))

cat_feature_names

array(['Item_Identifier_DRA12', 'Item_Identifier_DRA24',
       'Item_Identifier_DRA59', ..., 'Outlet_Type_Supermarket Type1',
       'Outlet_Type_Supermarket Type2', 'Outlet_Type_Supermarket Type3'],
      dtype=object)

In [ ]:
## get final col names which are num features + cat features
final_cols = numeric_select(X_train) + list(cat_feature_names)
final_cols

In [62]:
X_train_transformed = col_transformer.transform(X_train)

X_test_transformed = col_transformer.transform(X_test)

In [52]:
X_train_output = pd.DataFrame(X_train_transformed, columns = final_cols)

X_test_output = pd.DataFrame(X_test_transformed, columns = final_cols)

In [57]:
X_train_output.head()

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
0     0.817249        -0.712775  1.828109                   1.327849   
1     0.556340        -1.291052  0.603369                   1.327849   
2    -0.131512         1.813319  0.244541                   0.136187   
3    -1.169219        -1.004931 -0.952591                   0.732018   
4     1.528819        -0.965484 -0.336460                   0.493686   

    Outlet_Size  Outlet_Location_Type  Item_Identifier_DRA12  \
0  4.077854e-01              1.084948                    0.0   
1  4.077854e-01              1.084948                    0.0   
2  4.077854e-01             -1.384777                    0.0   
3 -1.291630e+00             -0.149914                    0.0   
4 -3.773459e-16             -0.149914                    0.0   

   Item_Identifier_DRA24  Item_Identifier_DRA59  Item_Identifier_DRB01  ...  \
0                    0.0                    0.0                    0.0  ...   
1                    0.0                    0.0                    0.0  ...   
2                    0.0                    0.0                    0.0  ...   
3                    0.0                    0.0                    0.0  ...   
4                    0.0                    0.0                    0.0  ...   

   Outlet_Identifier_OUT019  Outlet_Identifier_OUT027  \
0                       0.0                       0.0   
1                       0.0                       0.0   
2                       0.0                       0.0   
3                       0.0                       0.0   
4                       0.0                       0.0   

   Outlet_Identifier_OUT035  Outlet_Identifier_OUT045  \
0                       0.0                       0.0   
1                       0.0                       0.0   
2                       0.0                       0.0   
3                       1.0                       0.0   
4                       0.0                       1.0   

   Outlet_Identifier_OUT046  Outlet_Identifier_OUT049  \
0                       0.0                       0.0   
1                       0.0                       0.0   
2                       0.0                       1.0   
3                       0.0                       0.0   
4                       0.0                       0.0   

   Outlet_Type_Grocery Store  Outlet_Type_Supermarket Type1  \
0                        0.0                            0.0   
1                        0.0                            0.0   
2                        0.0                            1.0   
3                        0.0                            1.0   
4                        0.0                            1.0   

   Outlet_Type_Supermarket Type2  Outlet_Type_Supermarket Type3  
0                            1.0                            0.0  
1                            1.0                            0.0  
2                            0.0                            0.0  
3                            0.0                            0.0  
4                            0.0                            0.0  

[5 rows x 1588 columns]

In [59]:
X_test_output.head()

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
0  3.310089e-01        -0.776646 -0.998816                  -1.293807   
1 -1.179892e+00         0.100317 -1.585194                  -0.102145   
2  3.784469e-01        -0.482994 -1.595784                   0.136187   
3  4.213344e-16        -0.415440  0.506592                  -1.532139   
4 -6.426567e-01        -1.047426  0.886725                   0.732018   

   Outlet_Size  Outlet_Location_Type  Item_Identifier_DRA12  \
0     2.107200              1.084948                    0.0   
1    -1.291630             -1.384777                    0.0   
2     0.407785             -1.384777                    0.0   
3     0.407785              1.084948                    0.0   
4    -1.291630             -0.149914                    0.0   

   Item_Identifier_DRA24  Item_Identifier_DRA59  Item_Identifier_DRB01  ...  \
0                    0.0                    0.0                    0.0  ...   
1                    0.0                    0.0                    0.0  ...   
2                    0.0                    0.0                    0.0  ...   
3                    0.0                    0.0                    0.0  ...   
4                    0.0                    0.0                    0.0  ...   

   Outlet_Identifier_OUT019  Outlet_Identifier_OUT027  \
0                       0.0                       0.0   
1                       0.0                       0.0   
2                       0.0                       0.0   
3                       0.0                       1.0   
4                       0.0                       0.0   

   Outlet_Identifier_OUT035  Outlet_Identifier_OUT045  \
0                       0.0                       0.0   
1                       0.0                       0.0   
2                       0.0                       0.0   
3                       0.0                       0.0   
4                       1.0                       0.0   

   Outlet_Identifier_OUT046  Outlet_Identifier_OUT049  \
0                       0.0                       0.0   
1                       1.0                       0.0   
2                       0.0                       1.0   
3                       0.0                       0.0   
4                       0.0                       0.0   

   Outlet_Type_Grocery Store  Outlet_Type_Supermarket Type1  \
0                        0.0                            1.0   
1                        0.0                            1.0   
2                        0.0                            1.0   
3                        0.0                            0.0   
4                        0.0                            1.0   

   Outlet_Type_Supermarket Type2  Outlet_Type_Supermarket Type3  
0                            0.0                            0.0  
1                            0.0                            0.0  
2                            0.0                            0.0  
3                            0.0                            1.0  
4                            0.0                            0.0  

[5 rows x 1588 columns]

In [64]:
X_train_output.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6392 entries, 0 to 6391
Columns: 1588 entries, Item_Weight to Outlet_Type_Supermarket Type3
dtypes: float64(1588)
memory usage: 77.4 MB


In [65]:
X_test_output.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2131 entries, 0 to 2130
Columns: 1588 entries, Item_Weight to Outlet_Type_Supermarket Type3
dtypes: float64(1588)
memory usage: 25.8 MB


Build linear regression model to predict sales